src/neo-vm/ExecutionEngine.cs

```diff
@@ -42,6 +42,7 @@ public class ExecutionEngine : IDisposable
         public ExecutionContext EntryContext { get; private set; }
         public EvaluationStack ResultStack { get; }
         public VMState State { get; internal protected set; } = VMState.BREAK;
+        public FaultState FaultState { get; internal protected set; } = new FaultState { Rethrow = false };
```

Comment:
I think `FaultState` should be included in `TryContext`. Think about the following situation:

```c#
void A()
{
    try
    {
        throw new Exception("#1");
    }
    catch
    {
        B();
    }
}

void B()
{
    try
    {
        throw new Exception("#2");
    }
    catch
    {
    }
}
```

When `B()` returns, the fault state should be `#1`, not `#2`. Because `#2` has been handled.

Reply:
Not sure, I think both are ok. When we enter the catch of #1, vm has actually resumed.

Reply:
```c#
void A()
{
     try
    {
       B();
    }
    catch
    {
     }

}
void B()
{
    C();
}
void C()
{
   D();
}
void D()
{
   throw new exception("#1");
}
```
if fault state on trycontext, throw need to find trycontext.

how about let catch to clear fault state

```
void A()
{
    try
    {
        throw new Exception("#1");
    }
    catch
    {//clear fault state here.
        B();
    }
}

```

Reply:
But in `catch`, the contract needs to read the exception object. You need to offer the correct object.

Reply:
Maybe we need a `exception` field in `TryContext`, which means the catched exception.

Reply:
in c#，when catch，exception object had be push to stack default.
we can do same thing.

Reply:
Yes. It must be done.

Reply:
Done

Reply:
The exception should be pushed onto the evaluation stack.

Reply:
Then we need a new `VM.Types.Exception` type, or currently it's enough?

Reply:
I think we can't push the error onto the evaluation stack directly, like this example

```cs
try{
    //... 
   throw error
}finally{
    // ....blabla..
} // ENDF, we need to rethrow the error.

Reply:
You should push it at the beginning of the catch segment.

Reply:
So we still need the `exception` field of `FaultState`?

Reply:
Why? I don't think so.

Reply:
We need a way to save the error, before we meet the `catch` instruction. And user have already pushed the error onto the stack before the `throw`  instruction.

Reply:
The exception object is already in TryContext.

Reply:
```cs
try{// #1
     try{ // #2
            //... 
           throw error // 
      }finally{
             // ....blabla..
      } // ENDF, we need to rethrow the error.
}catch{  // push error onto the stack
  //
}
```
Where do we assign the error to the tryContext(`#1`)?  Or let's remove `CatchedException` of `TryContext` and keep the `Exception` of `FaultState`, then push the error at the begining of `catch` instruction.

Reply:
You don't need the `FaultState`.

https://github.com/neo-project/neo-vm/blob/745f51d4243de93783e4240b6ba39184f054d329/src/neo-vm/ExecutionEngine.cs#L306-L311

Throw the exception at here. (Not return)

https://github.com/neo-project/neo-vm/blob/745f51d4243de93783e4240b6ba39184f054d329/src/neo-vm/ExecutionEngine.cs#L1333-L1336

Then catch it and handle the exception at here.

Reply:
Done.

We have discussed this before, and it's better to use flag(FaultState) than try-catch in terms of performance.